cmd/updatetree/main_test.go

```diff
@@ -0,0 +1,178 @@
+	"github.com/google/trillian/testonly"
+	"github.com/google/trillian/util/flagsaver"
+)
+
+type testCase struct {
+	desc       string
+	setFlags   func()
+	updateErr  error
+	wantRPC    bool
+	updateTree *trillian.Tree
+	wantErr    bool
+	wantState  trillian.TreeState
+}
+
+func TestFreezeTree(t *testing.T) {
+	runTest(t, []*testCase{
+		{
+			// We don't set the treeID in runTest so this should fail.
+			desc:    "missingTreeID",
+			wantErr: true,
+		},
+		{
+			desc: "mandatoryOptsNotSet",
+			// Undo the flags set by runTest, so that mandatory options are no longer set.
+			setFlags: resetFlags,
+			wantErr:  true,
+		},
+		{
+			desc: "validUpdateFrozen",
+			setFlags: func() {
+				*treeID = 12345
+				*treeState = "FROZEN"
+			},
+			wantRPC: true,
+			updateTree: &trillian.Tree{
+				TreeId:    12345,
+				TreeState: trillian.TreeState_FROZEN,
+			},
+			wantState: trillian.TreeState_FROZEN,
+		},
+		{
+			desc: "updateInvalidState",
+			setFlags: func() {
+				*treeID = 12345
+				*treeState = "ITSCOLDOUTSIDE"
+			},
+			wantErr: true,
+		},
+		{
+			desc: "unknownTree",
+			setFlags: func() {
+				*treeID = 123456
+				*treeState = "FROZEN"
+			},
+			wantErr:   true,
+			wantRPC:   true,
+			updateErr: errors.New("unknown tree id"),
+		},
+		{
+			desc: "emptyAddr",
+			setFlags: func() {
+				*adminServerAddr = ""
+				*treeID = 12345
+				*treeState = "FROZEN"
+			},
+			wantErr: true,
+		},
+		{
+			desc: "updateErr",
+			setFlags: func() {
+				*treeID = 12345
+				*treeState = "FROZEN"
+			},
+			wantRPC:   true,
+			updateErr: errors.New("update tree failed"),
+			wantErr:   true,
+		},
+	})
+}
+
+// runTest executes the updateTree command against a fake TrillianAdminServer
+// for each of the provided tests, and checks that the tree in the request is
+// as expected, or an expected error occurs.
+// Prior to each test case, it:
+// 1. Resets all flags to their original values.
+// 2. Sets the adminServerAddr flag to point to the fake server.
+// 3. Calls the test's setFlags func (if provided) to allow it to change flags specific to the test.
+func runTest(t *testing.T, tests []*testCase) {
+	for _, tc := range tests {
+		t.Run(tc.desc, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			defer ctrl.Finish()
+
+			s, stopFakeServer, err := testonly.NewMockServer(ctrl)
+			if err != nil {
+				t.Fatalf("Error starting fake server: %v", err)
+			}
+			defer stopFakeServer()
+			defer flagsaver.Save().Restore()
+			*adminServerAddr = s.Addr
+			if tc.setFlags != nil {
+				tc.setFlags()
+			}
+
+			// We might not get as far as updating the tree on the admin server.
+			if tc.wantRPC {
+				call := s.Admin.EXPECT().UpdateTree(gomock.Any(), gomock.Any()).Return(tc.updateTree, tc.updateErr)
+				expectCalls(call, tc.updateErr)
+			}
+
+			ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+			defer cancel()
+			tree, err := updateTree(ctx)
+			switch hasErr := err != nil; {
+			case hasErr != tc.wantErr:
```

Comment:
The `switch`/`case` feels overkill here -- just `if hasErr := err != nil; hasErr !=tc.wantErr {` would work, wouldn't it?

Reply:
This has an extra `case` in the createtree version. I've gone with a simple `if` now.